src/base/DMRB_SPSC.hpp

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2008 Search Solution Corporation
+ * Copyright 2016 CUBRID Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+/*
+ * DMRB_SPSC.hpp
+ */
+
+#ifndef _DMRB_SPSC_HPP_
+#define _DMRB_SPSC_HPP_
+
+#ident "$Id$"
+
+#include "DMRB.hpp"
+
+#include <cstring>
+#include <cstddef>
+
+namespace cubbase
+{
+  /* single producer single consumer */
+  template <bool ThreadSafe>
+  class DMRB_SPSC : public DMRB<ThreadSafe>
+  {
+    public:
+      DMRB_SPSC (std::size_t capacity);
+      ~DMRB_SPSC ();
+
+      cubbase::span<std::byte> reserve (std::size_t length);
+      cubbase::span<std::byte> reserve_all ();
+      void commit (std::size_t length);
+
+      void consume (std::size_t length);
+
+      cubbase::span<const std::byte> peek () const;
+  };
+
+  template <bool T>
+  DMRB_SPSC<T>::DMRB_SPSC (std::size_t capacity) :
+    DMRB<T> (capacity)
+  {
+  }
+
+  template <bool T>
+  DMRB_SPSC<T>::~DMRB_SPSC ()
+  {
+  }
+
+  template <bool T>
+  cubbase::span<std::byte> DMRB_SPSC<T>::reserve (std::size_t length)
+  {
+    std::uint64_t head, tail;
+    std::byte *ptr;
+
+    head = DMRB<T>::value_load (this->m_head, std::memory_order_relaxed);
+    tail = DMRB<T>::value_load (this->m_tail, std::memory_order_acquire);
+
```

Comment:
```suggestion
assert(head >= tail);
assert(head - tail <= this->m_size);
```
NIT: 이런 assert문은 필요없을까요..?

Reply:
+ 다른 head, tail 부분도 동일합니다!

Reply:
다음 처리에서 함께 반영하겠습니다